sklearn/_build_utils/__init__.py

```diff
@@ -33,3 +35,36 @@ def atlas_not_found(blas_info_):
         cblas_libs = blas_info.pop('libraries', [])
 
     return cblas_libs, blas_info
+
+
+def get_cython_source(filename):
+    is_c_filename = filename.endswith('.c')
+    is_cpp_filename = filename.endswith('.cpp')
+
+    # files in src are .c and .cpp files that are not cython-generated
+    if 'src/' in filename and (is_c_filename or is_cpp_filename):
+        return filename
+    elif is_c_filename:
+        filename = filename[:-1]
+    elif is_cpp_filename:
+        filename = filename[:-3]
+    else:
+        raise ValueError('Only .c and .cpp files are supported. '
+                         'Got {0!r} instead'.format(filename))
+    return filename + 'pyx'
+
+
+def add_cython_extension(top_path, config, name, sources, **kwargs):
+    is_dev_version = not os.path.exists(os.path.join(top_path, 'PKG-INFO'))
+
+    if is_dev_version:
+        try:
+            from Cython.Build import cythonize
+        except ImportError:
+            raise ValueError('Please install cython in order '
+                             'to build a scikit-learn development version')
+
+        sources = [get_cython_source(filename) for filename in sources]
+    config.add_extension(name, sources, **kwargs)
+
+    config.ext_modules = cythonize(config.ext_modules)
```

Comment:
Hmmm I need to double-check this, maybe it should be inside an `if is_dev_version`